### PR TITLE
Removed `Azure::Core::Context::HasKey()` and add `TryGetValue()` instead

### PR DESCRIPTION
**Pull request by ahsonkhan**
~@vhvb1989 This change is blocked until after the retry policy implementation has been updated. Let me know once you have that PR available.~

Fixes https://github.com/Azure/azure-sdk-for-cpp/issues/2036

### CODE REVIEW

**Comment by vhvb1989** [comment]
I am confused abut removing this method without adding anything else like a `GetValueOrDefault(..)` to the Context.
Are we deciding to throw/abort if the value is not found in the context?

**Comment by ahsonkhan** [comment]
Waiting on https://github.com/Azure/azure-sdk-for-cpp/pull/2047 before this can get merged.

File: sdk/core/azure-core/test/ut/global_context.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/global_context.cpp:28]: @JeffreyRichter should `context.GetValue<...>()` throw if the key is not found or should it abort?

File: sdk/core/azure-core/inc/azure/core/context.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/context.hpp:151]: @CaseyCarter, in this case, would you recommend accepting an `outputValue` parameter as a reference (`T&`) or an explicit pointer (`T*`)?

I don't want someone to pass in a null pointer.

Does `T&&` makes sense here at all?

File: sdk/core/azure-core/test/ut/global_context.cpp
Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/global_context.cpp:28]: Also, do we expect the caller to know the type of the value corresponding to the key, to pass in as the template argument ahead of time?

Previously, calling `HasKey` didn't require knowledge of the `T`, but now the user needs to know and pass the right type to the `TryGetValue` method.

File: sdk/core/azure-core/inc/azure/core/context.hpp
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/context.hpp:151]: > Does T&& makes sense here at all?

It doesn't make sense to me.

Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/context.hpp:151]: > I don't want someone to pass in a null pointer.

It makes sense to me to allow passing null pointer. In which case, the user just wants to check if the key exists and the type matches.

**Review by Jinming-Hu — APPROVED** [review]
Also, by `TryGetValue`, you're assuming the value type is always copiable, which is not the case before this change. We only need it to be movable in the past.

**Comment by ahsonkhan** [comment]
> Also, by `TryGetValue`, you're assuming the value type is always copiable, which is not the case before this change. 

That's probably OK. Are there many basic/commonly used primitive/STL types that are movable but not copiable? Hopefully things like std::string, int, vector, and smart pointers continue to work.

> We only need it to be movable in the past.

@Jinming-Hu can you think of a way I can write `TryGetValue()` by using move semantics rather than copy?

File: sdk/core/azure-core/inc/azure/core/context.hpp
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/context.hpp:151]: > In which case, the user just wants to check if the key exists and the type matches.

That's an interesting idea. However, then the API behavior changes if the user passes in `null` vs non-null, which is confusing. 
- `null` ptr - always returns true or false, **false if type mismatch**
- non-null ptr - sometimes returns true or false, **abort if type mismatch** - what should TryGetValue do/return in this case?

I would prefer if the behavior was always consistent.

**Comment by Jinming-Hu** [comment]
> can you think of a way I can write `TryGetValue()` by using move semantics rather than copy?

There's small semantic difference here. `Get()` only returns a const reference to the value. But your impl of `TryGetValue` returns a new copy.

File: sdk/core/azure-core/inc/azure/core/context.hpp
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/context.hpp:151]: I'm thinking

* null ptr - returns true if key exists, else false. Type mismatch will throw or abort.

Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/context.hpp:151]: I prefer `T&`

Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/context.hpp:151]: That behavior is equivalent to the `HasKey` method, which we are removing because:
- there is no real scenario for it, where the caller wouldn't then try to get the actual value after that returns true anyway
- Using HasKey then GetValue encourages a poor coding pattern where we do two look-ups within the context tree

Can you think of a real-world use case or scenario where the user would want to pass in a null ptr to `TryGetValue` (or previously call `HasKey`) that isn't followed by making another call to get the value?

**Comment by ahsonkhan** [comment]
> But your impl of `TryGetValue` returns a new copy.

Can you help me understand how is it returning a new copy? I am assigning the underlying value to the reference that the caller passed in as a parameter. Why is that different from what `Get()` is doing? The implementation of the two methods are almost identical. Does it have something to do with the **const ref** return value of `Get()`?

```C++
 // This takes the context value under the covers and assigns it to `val`
std::string val;
context.TryGetValue<std::string>("Key", val);
// Isn't the similar assignment happening at the caller-end, even with Get?
val = context.Get<std::string>("Key");
```

**Comment by Jinming-Hu** [comment]
> > But your impl of `TryGetValue` returns a new copy.
> 
> Can you help me understand how is it returning a new copy? I am assigning the underlying value to the reference that the caller passed in as a parameter. Why is that different from what `Get()` is doing? The implementation of the two methods are almost identical. Does it have something to do with the **const ref** return value of `Get()`?
> 
> ```c++
>  // This takes the context value under the covers and assigns it to `val`
> std::string val;
> context.TryGetValue<std::string>("Key", val);
> // Isn't the similar assignment happening at the caller-end, even with Get?
> val = context.Get<std::string>("Key");
> ```

There's no difference in your use case. But there's a difference in the following code.

```
const std::string& val = context.Get<std::string>("key");  // There's no copy

std::string val2;
context.TryGetValue<std::string>("Key", val2);  // There's a copy
```

Which, as I already said, is fine by me.

File: sdk/core/azure-core/inc/azure/core/context.hpp
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/context.hpp:151]: For example, the size of value is big and copy can be expensive. The key may not exist. In the past user can do

```C++
if (context.HasKey("key")) {
    const ValueType& val = context.GetValue("key");
}
```

Now there's no way to do that.

An option is to throw exception from `GetValue` if key doesn't exist.

Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/context.hpp]: well @Jinming-Hu we would use a raw-pointer or smart-ptr in order to avoid the copy and just access to it
like
```cpp
BigStructure value;
context c = otherContext.WithValue(key, &value);

// Then to read it from context
auto ref = c.TryGetValue<BigStructure*>(key);
```

Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/context.hpp]: @vhvb1989 Yeah, that's an option.

File: sdk/core/azure-core/inc/azure/core/context.hpp
Comment (antkmsft) [inline sdk/core/azure-core/inc/azure/core/context.hpp:197]: Write a common function and invoke it from two places:
```
private: template <class T> bool TryGetValue(Key const& key, T const** outputValue) const
// does everything TryGetValue does, but returns pointer to found value

public:
template <class T> bool TryGetValue(Key const& key, T& outputValue) const
{
    T const* valuePtr = nullptr;
    if (this->TryGetValue<T>(key, &valuePtr))
    {
        outputValue = *valuePtr;
        return true;
    }
    return false;
}

template <class T> T const& GetValue(Key const& key) const
{
    T const* valuePtr = nullptr;
    if (this->TryGetValue<T>(key, &valuePtr))
    {
        return *valuePtr;
    }
    std::abort(); // grrr, #1986
}
```

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/context.hpp:197]: I'd also verify if GetValue() is needed at all, or do we, in SDK, always check.

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (antkmsft) [inline sdk/core/azure-core/src/http/retry_policy.cpp:96]:
```suggestion
  int32_t number = -1;
```

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (antkmsft) [inline sdk/core/azure-core/src/http/retry_policy.cpp:97]:
```suggestion
  if (!context.TryGetValue<int32_t>(RetryKey, number))
```

File: sdk/core/azure-core/CHANGELOG.md
Comment (antkmsft) [inline sdk/core/azure-core/CHANGELOG.md:27]: Add "- Removed `HasKey()`." to "Changes to `Azure::Core::Context`:" section (line 22)

File: sdk/core/azure-core/inc/azure/core/context.hpp
Comment (antkmsft) [inline sdk/core/azure-core/inc/azure/core/context.hpp:167]: backtick `true` and `false`, tag `outputValue` (is it "`\p`"?)

File: sdk/core/azure-core/inc/azure/core/context.hpp
Comment (antkmsft) [inline sdk/core/azure-core/inc/azure/core/context.hpp:164]: tag `key` parameter

File: sdk/core/azure-core/inc/azure/core/context.hpp
Comment (antkmsft) [inline sdk/core/azure-core/inc/azure/core/context.hpp:194]: It should've been "A value associated with the *key*  ... if a specific *key* can't be found", but second part of the doc is inaccurate. We won't return an empty value, we would UB, or, practically, `std::abort()`.

File: sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp
Comment (antkmsft) [inline sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp:16]: The way it is used, let's guarantee (in the docs) that the value will be untouched if the key was not found.

Comment (antkmsft) [reply sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp:16]: This also begs the question if we use non-`Try` version of this method at all.

**Review by antkmsft — CHANGES_REQUESTED** [review]
Other than that, LGTM

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (Jinming-Hu) [reply sdk/core/azure-core/src/http/retry_policy.cpp:96]: I prefer `int`.

File: sdk/core/azure-core/inc/azure/core/context.hpp
Comment (JeffreyRichter) [inline sdk/core/azure-core/inc/azure/core/context.hpp:197]: I would remove this function now.
The common usage is for a policy to look for a key and for it NOT to be there; this is why TryGetValue returns false. But, because it is common for the key to not be there, GetValue should never be called.

**Review by RickWinter — DISMISSED** [review]
Please hold off on this change

**Comment by ahsonkhan** [comment]
> Please hold off on this change

@RickWinter Can you please provide more details. Until when? And why?

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (antkmsft) [reply sdk/core/azure-core/src/http/retry_policy.cpp:96]: I am fine with `int`, however `RetryOptions` has it as `int32_t`, they should be in sync.

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (antkmsft) [reply sdk/core/azure-core/src/http/retry_policy.cpp:97]: I am fine with `int`, however RetryOptions has it as `int32_t`, they should be in sync.

File: sdk/core/azure-core/CHANGELOG.md
Comment (antkmsft) [inline sdk/core/azure-core/CHANGELOG.md:23]: "Removed `Get()` and `HasKey()` in favor of a new method `TryGetValue()`"? So, it is actionable as you read it, what to replace it with.

File: sdk/core/azure-core/test/ut/context.cpp
Comment (antkmsft) [inline sdk/core/azure-core/test/ut/context.cpp:22]: Btw, it is possible that we can write just `TryGetValue(...)` or `TryGetValue<>(...)`, and a compiler will deduce the value of `<T>`.

File: sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp
Comment (antkmsft) [inline sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp:16]: If it is possible to invoke without explicitly putting `<std::shared_ptr<bool>>`, it is better

File: sdk/core/azure-core/test/ut/global_context.cpp
Comment (antkmsft) [inline sdk/core/azure-core/test/ut/global_context.cpp:28]: Test that the value hasn't changed when `false` is returned.

File: sdk/core/azure-core/test/ut/context.cpp
Comment (antkmsft) [inline sdk/core/azure-core/test/ut/context.cpp:145]: verify that the value hasn't changed when `false` is returned

File: sdk/core/azure-core/test/ut/context.cpp
Comment (antkmsft) [inline sdk/core/azure-core/test/ut/context.cpp:137]: Verify that the value hasn't changed when `false` is returned

File: sdk/core/azure-core/test/ut/context.cpp
Comment (antkmsft) [inline sdk/core/azure-core/test/ut/context.cpp:122]: Verify that the value hasn't changed when `false` is returned

File: sdk/core/azure-core/test/ut/context.cpp
Comment (antkmsft) [inline sdk/core/azure-core/test/ut/context.cpp:105]: Verify that the value hasn't changed when `false` is returned

File: sdk/core/azure-core/test/ut/context.cpp
Comment (antkmsft) [inline sdk/core/azure-core/test/ut/context.cpp:60]: nit: if you cound update `EXPECT_TRUE(x == y)` to `EXPECT_EQ(x, y)`, that'd be great. Logs/test failure messages are more informative this way.

File: sdk/core/azure-core/test/ut/context.cpp
Comment (antkmsft) [reply sdk/core/azure-core/test/ut/context.cpp:22]: Verify that the value hasn't changed when `false` is returned.

**Review by antkmsft — APPROVED** [review]

File: sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp
Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp:16]: From the existing code I read, it wasn't possible with `GetValue`, so why would it be possible now with `TryGetValue`?

I will try it and see.

File: sdk/core/azure-core/test/ut/context.cpp
Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/context.cpp:145]: I don't think we need to test that everywhere. Having the assert in a couple places would be sufficient.

File: sdk/core/azure-core/test/ut/context.cpp
Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/context.cpp:137]: Will skip adding it everywhere, like here and below. It doesn't add value to duplicate the testing.

File: sdk/core/azure-core/test/ut/context.cpp
Comment (antkmsft) [reply sdk/core/azure-core/test/ut/context.cpp:145]: That's also fine. I'd rather add a few of excessive lines, but one dedicated test case testing that the value hasn't been modified is also fine. Just make sure you assign it some interesting initial value that is unlikely to get by a chance (i.e. `1234` instead of `0` or `-1`).

File: sdk/core/azure-core/test/ut/context.cpp
Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/context.cpp:60]: Unrelated to this PR and is an existing issue throughout the SDK tests. Leaving as is for now. https://github.com/Azure/azure-sdk-for-cpp/issues/2057

File: sdk/core/azure-core/test/ut/context.cpp
Comment (antkmsft) [inline sdk/core/azure-core/test/ut/context.cpp:25]: I'd add a test case for `std::shared_ptr<something>` - the type that Storage stores there.

Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/context.cpp:25]: Can you please provide an example code snippet of how to write such a test.

File: sdk/core/azure-core/test/ut/context.cpp
Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/context.cpp:22]: > Btw, it is possible that we can write just `TryGetValue(...)` or `TryGetValue<>(...)`, and a compiler will deduce the value of `<T>`.

Verified that's possible. Good call. I will leave it as explicit in tests though.

File: sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp
Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp:16]: It works without it. Leaving it as is, since the existing code was explicit.

File: sdk/core/azure-core/test/ut/context.cpp
Comment (antkmsft) [reply sdk/core/azure-core/test/ut/context.cpp:25]: Something like this? :

TEST(...)
{
    class S {
    private:
        int* m_instanceCount;
    public:
        S(int* instanceCount) : m_instanceCount(instanceCount) { ++(*m_instanceCount); }
        ~S() { --(*m_instanceCount); }
    };

    int instanceCount = 0;
    {
        auto sharedPtr = std::make_shared<S>(&instanceCount);
        EXPECT_EQ(sharedPtr.use_count(), 1);

        context = context.WithValue(key, sharedPtr);
        EXPECT_EQ(sharedPtr.use_count(), 2);

        std::shared_ptr<S> foundPtr;
        EXPECT_TRUE(context.TryGetValue(key, foundPtr));
        EXPECT_EQ(foundPtr.get(), sharedPtr.get());
        EXPECT_EQ(*instanceCount, 1);
        EXPECT_EQ(sharedPtr.use_count(), 1);
    }

   // Verify that context calls the destructor of shared_ptr it is holding
   EXPECT_EQ(*instanceCount, 0);
}
```

Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/context.cpp:25]: storage is using shared ptr of `bool`, so I will use that, but also see if I can add the complex class scenario like you showed.

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (antkmsft) [inline sdk/core/azure-core/src/http/retry_policy.cpp:105]: With the new guarantee that we won't touch the value, we don't need any more assignments to the pointer. it is already associated with an int that equals to -1.
BTW, I'd change the value we store by `RetryKey` to "`int32 const*`" - we don't expect calling code to modify it.

File: sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp
Comment (antkmsft) [inline sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp:16]: Nice! - so we can get the value to get deducted by compiler. Should we remove the rest of the type specifications?

Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp:16]: We can, but I will leave that as a separate change, and to each area owner. Some folks like being explicit. I don't mind being explicit either.

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/retry_policy.cpp:105]: > BTW, I'd change the value we store by `RetryKey` to "`int32 const*`" - we don't expect calling code to modify it.

Unrelated change. I will leave it to @vhvb1989 to do so separately. If you want to do it @antkmsft, go for it :)

File: sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp
Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp]: @Jinming-Hu , in storage code, which do you prefer? Explicitly stating the `<T>` or leaving it out and it being inferred?

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/src/storage_switch_to_secondary_policy.cpp]: Leave it out

**Comment by ahsonkhan** [comment]
@RickWinter please let me know if there is still anything else blocking. I will address those separately.

**Pull request merged**
